lib/domain/validation/authenticator_validator.dart

```diff
@@ -0,0 +1,116 @@
+    List<String> authenticatorAllowList,
+  );
+
+  Set<Authenticator> validateForAuthentication(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  );
+}
+
+@Injectable(as: AuthenticatorValidator)
+class AuthenticatorValidatorImpl implements AuthenticatorValidator {
+  @override
+  Future<Set<Authenticator>> validateForRegistration(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    Set<Authenticator> allowedAuthenticators = _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    );
+
+    return whereAsync(allowedAuthenticators, (authenticator) async {
+      // Do not display:
+      //  - policy non-compliant authenticators (this includes already registered authenticators)
+      //  - not hardware supported authenticators
+      //  - prefer Biometrics authenticator on Android
+      return authenticator.isSupportedByHardware &&
+          await context.isPolicyCompliant(authenticator.aaid) &&
+          await _filterAndroidFingerprintIfNecessary(context, authenticator);
+    });
+  }
+
+  @override
+  Set<Authenticator> validateForAuthentication(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    return _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    ).where((authenticator) {
+      // Do not display:
+      //   - non-registered authenticators
+      //   - not hardware supported authenticators
+      return authenticator.registration
+              .isRegistered(context.account.username) &&
+          authenticator.isSupportedByHardware;
+    }).toSet();
+  }
+
+  Set<Authenticator> _allowedAuthenticators(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    return context.authenticators
+        .where((a) => authenticatorAllowList.contains(a.aaid))
+        .toSet();
+  }
+
+  Future<bool> _filterAndroidFingerprintIfNecessary(
+    AuthenticatorSelectionContext context,
+    Authenticator authenticator,
+  ) async {
+    if (Platform.isIOS) {
+      return Future.value(true);
+    }
+
+    final authenticators = context.authenticators;
+    final isBiometricRegistered = await anyAsync(
+      authenticators,
+      (Authenticator a) => Future.value(
+        a.aaid.isBiometric &&
+            a.registration.isRegistered(context.account.username),
+      ),
+    );
+    final canRegisterBiometric =
+        await anyAsync(authenticators, (Authenticator a) async {
+      return Future.value(
+        a.aaid.isBiometric &&
+            await context.isPolicyCompliant(a.aaid) &&
+            a.isSupportedByHardware,
+      );
+    });
+    final canRegisterFingerprint =
+        await anyAsync(authenticators, (Authenticator a) async {
+      return Future.value(
+        a.aaid.isFingerprint &&
+            await context.isPolicyCompliant(a.aaid) &&
+            a.isSupportedByHardware,
+      );
+    });
```

Comment:
🔧 I think Fingerprint authenticator is always supported by hardware so combining this with my other comment about policy compliance being already checked before, we don't eed this `canRegisterFingerprint` check at all.